### PR TITLE
Bugfix/fix multiserver session state updates

### PR DESCRIPTION
**Pull request by ieva-improbable**
We are currently not doing any authority checks for updating the two session components, this PR makes sure that only the authoritative server makes the updates.

**Note:** we need to merge release into master (see [this PR](https://github.com/spatialos/UnrealGDKExampleProject/pull/4)) before merging this change in, as it will otherwise come with a bunch of unrelated commits.

### CODE REVIEW

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp [hunk 1]
```diff
@@ -44,6 +45,15 @@ void AGDKSessionGameState::RemovePlayerState(APlayerState* PlayerState)
 
 void AGDKSessionGameState::OnRep_SessionProgress()
 {
+	USpatialNetDriver* SpatialNetDriver = Cast<USpatialNetDriver>(GetWorld()->GetNetDriver());
+	bool AuthoritativeOverSessionEntity = SpatialNetDriver->StaticComponentView->HasAuthority(SessionEntityId, SessionComponentId);
```
Comment (m-samiec) [inline Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]:
```suggestion
	bool bAuthoritativeOverSessionEntity = SpatialNetDriver->StaticComponentView->HasAuthority(SessionEntityId, SessionComponentId);
```

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp [hunk 2]
```diff
@@ -44,6 +45,15 @@ void AGDKSessionGameState::RemovePlayerState(APlayerState* PlayerState)
...
+
+	if (AuthoritativeOverSessionEntity)
+	{
+		// There's an offset of 1 between the corresponding states of session progress and session state.
+		int SessionState = (int)SessionProgress + 1;
+		SendStateUpdate(SessionState);
```
Comment (m-samiec) [inline Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]: I realise you haven't changed this, but could we switch SendStateUpdate to internally use SessionProgress (and use a static_cast<> to cast it to and int), and offset it by 1 in the component update?

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp [hunk 3]
```diff
@@ -64,29 +74,43 @@ void AGDKSessionGameState::BeginTimer()
 
 void AGDKSessionGameState::TickGameTimer()
 {
-	if (GetNetMode() != NM_Client)
+	bool AuthoritativeOverSessionProgress = Role == ROLE_Authority;
```
Comment (m-samiec) [inline Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]:
```suggestion
	const bool bAuthoritativeOverSessionProgress = HasAuthority();
```

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp [hunk 4]
```diff
@@ -64,29 +74,43 @@ void AGDKSessionGameState::BeginTimer()
...
+
+	if (GetNetMode() != NM_Client && AuthoritativeOverSessionProgress)
 	{
 		SessionTimer--;
 
+		USpatialNetDriver* SpatialNetDriver = Cast<USpatialNetDriver>(GetWorld()->GetNetDriver());
+		bool AuthoritativeOverSessionEntity = SpatialNetDriver->StaticComponentView->HasAuthority(SessionEntityId, SessionComponentId);
```
Comment (m-samiec) [inline Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]:
```suggestion
		bool bAuthoritativeOverSessionEntity = SpatialNetDriver->StaticComponentView->HasAuthority(SessionEntityId, SessionComponentId);
```

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp [hunk 5]
```diff
@@ -6,6 +6,7 @@
 #include "UnrealNetwork.h"
 #include "SpatialNetDriver.h"
 #include "SpatialWorkerConnection.h"
+#include "Interop/SpatialStaticComponentView.h"
 #include "c_worker.h"
```
Comment (m-samiec) [inline Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]: Driveby - can we switch these to use the full path #include <WorkerSDK/improbable/...>?

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp (hunk 2 above)
Comment (ieva-improbable) [reply Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]: Is it better now?

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp [hunk 6]
```diff
@@ -64,46 +72,63 @@ void AGDKSessionGameState::BeginTimer()
...
+void AGDKSessionGameState::SendStateUpdate(EGDKSessionProgress SessionProgress)
 {
 	if (!GetWorld()->GetNetDriver() || !GetWorld()->GetNetDriver()->IsA<USpatialNetDriver>())
 	{
 		return;
 	}
 
-	Worker_EntityId target_entity_id = 39;
+	// There's an offset of 1 between the corresponding states of session progress and session state.
+	int SessionState = (int)SessionProgress + 1;
```
Comment (m-samiec) [inline Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]:
```suggestion
	int SessionState = static_cast<int>(SessionProgress) + 1;
```

File: Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp [hunk 7]
```diff
@@ -64,46 +72,63 @@ void AGDKSessionGameState::BeginTimer()
...
+			if (bAuthoritativeOverSessionEntity)
+			{
+				SendStateUpdate(SessionProgress);
+			}
 		}
 	}
 }
 
-void AGDKSessionGameState::SendStateUpdate(int NewState)
+void AGDKSessionGameState::SendStateUpdate(EGDKSessionProgress SessionProgress)
```
Comment (m-samiec) [inline Game/Source/GDKShooter/Private/Game/GDKSessionGameState.cpp]: This is hiding a member variable name and will cause a warning, perhaps rename to `InSessionProgress`.
